sklearn/tree/tests/test_tree.py

```diff
@@ -2225,3 +2227,168 @@ def get_joblib_pickle_non_native_endianness():
+
+    n_classes = np.array([0, 1], dtype=expected_dtype)
+    for dt in allowed_dtypes:
+        check_n_classes(n_classes.astype(dt), expected_dtype)
+
+    with pytest.raises(ValueError, match="Wrong dimensions.+n_classes"):
+        wrong_dim_n_classes = np.array([[0, 1]], dtype=expected_dtype)
+        check_n_classes(wrong_dim_n_classes, expected_dtype)
+
+    with pytest.raises(ValueError, match="n_classes.+incompatible dtype"):
+        wrong_dtype_n_classes = n_classes.astype(np.float64)
+        check_n_classes(wrong_dtype_n_classes, expected_dtype)
+
+
+def test_check_value_ndarray():
+    expected_dtype = np.dtype(np.float64)
+    expected_shape = (5, 1, 2)
+    value_ndarray = np.zeros((5, 1, 2), dtype=expected_dtype)
+
+    allowed_dtypes = [expected_dtype, expected_dtype.newbyteorder()]
+
+    for dt in allowed_dtypes:
+        check_value_ndarray(
+            value_ndarray, expected_dtype=dt, expected_shape=expected_shape
+        )
+
+    with pytest.raises(ValueError, match="Wrong shape.+value array"):
+        check_value_ndarray(
+            value_ndarray, expected_dtype=expected_dtype, expected_shape=(1, 2)
+        )
+
+    for problematic_arr in [value_ndarray[:, :, :1], np.asfortranarray(value_ndarray)]:
+        with pytest.raises(ValueError, match="value array.+C-contiguous"):
+            check_value_ndarray(
+                problematic_arr,
+                expected_dtype=expected_dtype,
+                expected_shape=problematic_arr.shape,
+            )
+
+    with pytest.raises(ValueError, match="value array.+incompatible dtype"):
+        check_value_ndarray(
+            value_ndarray.astype(np.float32),
+            expected_dtype=expected_dtype,
+            expected_shape=expected_shape,
+        )
+
+
+def test_check_node_ndarray():
+    expected_dtype = NODE_DTYPE
+
+    node_ndarray = np.zeros((5,), dtype=expected_dtype)
+
+    valid_node_ndarrays = [
+        node_ndarray,
+        get_different_bitness_node_ndarray(node_ndarray),
+    ]
+    valid_node_ndarrays += [
+        arr.astype(arr.dtype.newbyteorder()) for arr in valid_node_ndarrays
+    ]
+    # TODO add test array with padding?
+
+    for arr in valid_node_ndarrays:
+        check_node_ndarray(node_ndarray, expected_dtype=expected_dtype)
+
+    with pytest.raises(ValueError, match="Wrong dimensions.+node array"):
+        problematic_node_ndarray = np.zeros((5, 2), dtype=expected_dtype)
+        check_node_ndarray(problematic_node_ndarray, expected_dtype=expected_dtype)
+
+        with pytest.raises(ValueError, match="node array.+C-contiguous"):
+            problematic_node_ndarray = node_ndarray[::2]
+            check_node_ndarray(problematic_node_ndarray, expected_dtype=expected_dtype)
+
+    # TODO incompatible dtypes e.g. change one float64 array by float32 or
+    # change one indexing dtype by float64
+    # with pytest.raises(ValueError, match="node array.+incompatible dtype"):
+    #     check_node_ndarray()
```

Comment:
```suggestion
    # The following NODE_DTYPE can never be valid on any platform because the
    # threshold field is float32 instead of the expected float64:
    INVALID_NODE_DTYPE = np.dtype(
        [
            ("left_child", "<i8"),
            ("right_child", "<i8"),
            ("feature", "<i8"),
            ("threshold", "<f4"),
            ("impurity", "<f8"),
            ("n_node_samples", "<i8"),
            ("weighted_n_node_samples", "<f8"),
        ]
    )
    with pytest.raises(ValueError, match="node array.+incompatible dtype"):
        check_node_ndarray(np.zeros(3, dtype=INVALID_NODE_DTYPE))
```